@Where on entity class level leads to boot exception

It seems the fix for #944 introduced logic that fails the boot when a `@Where` annotation is used on the entity class level. The exception is:

```
Caused by: org.hibernate.QueryException: could not resolve property: class of: my.domain.entity.tol.TolEntry
	at org.hibernate.persister.entity.AbstractPropertyMapping.propertyException(AbstractPropertyMapping.java:73) ~[hibernate-core-5.4.21.Final.jar:5.4.21.Final]
	at org.hibernate.persister.entity.AbstractPropertyMapping.toType(AbstractPropertyMapping.java:67) ~[hibernate-core-5.4.21.Final.jar:5.4.21.Final]
	at org.hibernate.persister.entity.AbstractEntityPersister.getSubclassPropertyTableNumber(AbstractEntityPersister.java:2075) ~[hibernate-core-5.4.21.Final.jar:5.4.21.Final]
	at org.hibernate.persister.entity.BasicEntityPropertyMapping.toColumns(BasicEntityPropertyMapping.java:46) ~[hibernate-core-5.4.21.Final.jar:5.4.21.Final]
	at org.hibernate.persister.entity.AbstractEntityPersister.toColumns(AbstractEntityPersister.java:2017) ~[hibernate-core-5.4.21.Final.jar:5.4.21.Final]
	at com.blazebit.persistence.integration.hibernate.base.HibernateJpaProvider.hasJoinCondition(HibernateJpaProvider.java:1419) ~[blaze-persistence-integration-hibernate-base-1.5.1.jar:1.5.1]
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,7 @@ Not yet released
 ### Bug fixes
 
 * Fix Hibernate 5.4.19+ integration issue
+* Fix boot exception when using Hibernate `@Where` on entity level
 
 ### Backwards-incompatible changes
 

--- a/core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/Issue1167Test.java
+++ b/core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/Issue1167Test.java
@@ -1,0 +1,54 @@
+/*
+ * Copyright 2014 - 2020 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.testsuite;
+
+import org.hibernate.annotations.Where;
+import org.junit.Test;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.ManyToOne;
+
+/**
+ * @author Christian Beikov
+ * @since 1.6.0
+ */
+public class Issue1167Test extends AbstractCoreTest {
+
+    @Override
+    protected Class<?>[] getEntityClasses() {
+        return new Class<?>[]{ BasicEntity.class };
+    }
+
+    @Entity(name = "BasicEntity")
+    @Where(clause = "deleted_char <> 't'")
+    public static class BasicEntity {
+        @Id
+        Long id;
+        @Column(name = "deleted_char", length = 1, nullable = false)
+        char deleted = 'f';
+        @ManyToOne
+        BasicEntity parent;
+    }
+
+    @Test
+    public void test1() {
+        cbf.create(em, BasicEntity.class).getResultList();
+    }
+
+}

--- a/integration/hibernate-base/src/main/java/com/blazebit/persistence/integration/hibernate/base/HibernateJpaProvider.java
+++ b/integration/hibernate-base/src/main/java/com/blazebit/persistence/integration/hibernate/base/HibernateJpaProvider.java
@@ -1410,16 +1410,10 @@ public class HibernateJpaProvider implements JpaProvider {
                 AbstractEntityPersister elementPersister = (AbstractEntityPersister) entityPersisters.get(((org.hibernate.type.EntityType) propertyType).getAssociatedEntityName());
                 String filterFragment = elementPersister.filterFragment("x", Collections.emptyMap());
                 if (elementPersister instanceof SingleTableEntityPersister) {
-                    if (!filterFragment.isEmpty()) {
-                        StringBuilder sb = new StringBuilder();
-                        sb.append(" and ");
-                        String discriminatorColumnName = elementPersister.getDiscriminatorColumnName();
+                    String discriminatorColumnName;
+                    if (!filterFragment.isEmpty() && (discriminatorColumnName = elementPersister.getDiscriminatorColumnName()) != null) {
                         InFragment frag = new InFragment();
-                        if (discriminatorColumnName == null) {
-                            frag.setFormula("x", elementPersister.toColumns("x", "class")[0]);
-                        } else {
-                            frag.setColumn("x", discriminatorColumnName);
-                        }
+                        frag.setColumn("x", discriminatorColumnName);
 
                         String[] subclasses = ((SingleTableEntityPersister) elementPersister).getSubclassClosure();
                         for (int i = 0; i < subclasses.length; i++) {

